### ISSUE TITLE
backup: delete tables in an optimal order

### ISSUE THREAD

**Issue by mmatczuk** [scylladb/scylla-manager#2815]
In e664a567f7aa813be288f78629b98367e59ee8d8 the table upload order is calculated based on size - biggest tables get priority.
This is not the best strategy in all cases, the problem with it is that it does not take into account data deduplication.
In production setting we may find tables such that uploading few Megabytes would free Gigabytes.

Let
 - `t_0, t_1, ..., t_n-1` be table snapshots we need to upload
 - `s(t_x)` be disk size of table x
 - `d(t_x)` be size of duplicate files (no need to upload them) of table x
 - `u(t_x) = s(t_x) - d(t_x)` be size of files to upload of table x

Then the loss function is total time each snapshot byte stayed on disk.

```
loss = SUM[i in [0, n)] u(t_i) * SUM [j in [i,n)] d(t_j)
```

**Comment by mmatczuk** [comment]
I'm not sure if there is a smart exact algorithm we could use but we can approximate it https://resources.wolframcloud.com/FunctionRepository/resources/MaximizeOverPermutations

**Comment by mmatczuk** [comment]
Monte Carlo crashes the optimization problem

![image](https://user-images.githubusercontent.com/1617930/124281746-541a3a80-db4a-11eb-8e67-dbd5a0cf7c19.png)

However based on my data from production with one huge table and a bunch of others with different degree of deduplication it comes to a conclusion to put the biggest table last.

**Comment by mmatczuk** [comment]
We shall estimate duplication level by comparing files dates vs previous backup timestamp.

**Comment by mmatczuk** [comment]
Below is the code to optimize the loss function described above.

```go
package experiment

import (
	"fmt"
	"math"
	"math/rand"
	"testing"

	"github.com/rclone/rclone/fs"
	"github.com/scylladb/scylla-manager/pkg/util/timeutc"
	"gonum.org/v1/plot"
	"gonum.org/v1/plot/plotter"
	"gonum.org/v1/plot/plotutil"
	"gonum.org/v1/plot/vg"
)

// a │     100% │    978MiB │    978MiB │       905MiB │     0B │ 30 Jun 21 11:29:31 UTC │ 30 Jun 21 11:29:35 UTC │
// b │     100% │      4MiB │      4MiB │         4MiB │     0B │ 30 Jun 21 11:29:31 UTC │ 30 Jun 21 11:29:31 UTC │
// c │     100% │   1.21GiB │   1.21GiB │      1.11GiB │     0B │ 30 Jun 21 11:29:15 UTC │ 30 Jun 21 11:29:22 UTC │
// d │     100% │    198MiB │    198MiB │       198MiB │     0B │ 30 Jun 21 11:27:46 UTC │ 30 Jun 21 11:27:46 UTC │
// e │     100% │   2.79GiB │   2.79GiB │      2.61GiB │     0B │ 30 Jun 21 11:27:38 UTC │ 30 Jun 21 11:27:46 UTC │
// f │     100% │    182MiB │    182MiB │           0B │     0B │ 30 Jun 21 11:27:31 UTC │ 30 Jun 21 11:27:38 UTC │
// g │     100% │      1MiB │      1MiB │           0B │     0B │ 30 Jun 21 11:27:28 UTC │ 30 Jun 21 11:27:31 UTC │
// h │     100% │    205MiB │    205MiB │       127MiB │     0B │ 30 Jun 21 11:27:26 UTC │ 30 Jun 21 11:27:28 UTC │
// i │     100% │   1.21GiB │   1.21GiB │      1.19GiB │     0B │ 30 Jun 21 11:27:24 UTC │ 30 Jun 21 11:27:26 UTC │
// j │     100% │     25MiB │     25MiB │        25MiB │     0B │ 30 Jun 21 11:27:16 UTC │ 30 Jun 21 11:27:16 UTC │
// k │     100% │  21.22GiB │  21.22GiB │     20.95GiB │     0B │ 30 Jun 21 09:27:20 UTC │ 30 Jun 21 09:27:29 UTC │
// l │     100% │  32.75GiB │  32.75GiB │     32.23GiB │     0B │ 30 Jun 21 09:27:09 UTC │ 30 Jun 21 09:27:20 UTC │
// m │     100% │     18MiB │     18MiB │        18MiB │     0B │ 30 Jun 21 11:27:46 UTC │ 30 Jun 21 11:27:46 UTC │
// n │     100% │ 473.30GiB │ 473.30GiB │    437.18GiB │     0B │ 30 Jun 21 09:18:13 UTC │ 30 Jun 21 09:27:09 UTC │
// o │     100% │     27MiB │     27MiB │           0B │     0B │ 30 Jun 21 09:18:10 UTC │ 30 Jun 21 09:18:13 UTC │
// p │     100% │  49.09GiB │  49.09GiB │     48.75GiB │     0B │ 30 Jun 21 11:27:55 UTC │ 30 Jun 21 11:28:04 UTC │
// q │     100% │  18.85GiB │  18.85GiB │     18.69GiB │     0B │ 30 Jun 21 11:27:20 UTC │ 30 Jun 21 11:27:24 UTC │
// r │     100% │   4.52GiB │   4.52GiB │      4.10GiB │     0B │ 30 Jun 21 11:29:22 UTC │ 30 Jun 21 11:29:31 UTC │
// s │     100% │ 152.12GiB │ 152.12GiB │    147.25GiB │     0B │ 30 Jun 21 11:28:04 UTC │ 30 Jun 21 11:29:15 UTC │
// t │     100% │  2.280TiB │  2.280TiB │     1.653TiB │     0B │ 30 Jun 21 09:27:29 UTC │ 30 Jun 21 11:27:15 UTC │
// x │     100% │   4.68GiB │   4.68GiB │      4.18GiB │     0B │ 30 Jun 21 11:27:46 UTC │ 30 Jun 21 11:27:55 UTC │
// y │     100% │   1.16GiB │   1.16GiB │      1.14GiB │     0B │ 30 Jun 21 11:27:16 UTC │ 30 Jun 21 11:27:20 UTC │

var table = []struct {
	Table string
	Size  float64
	Dup   float64
}{
	{"a", 978, 905},
	{"b", 4, 4},
	{"c", 1.21 * 1024, 1.11 * 1024},
	{"d", 198, 198},
	{"e", 2.79 * 1024, 2.61 * 1024},
	{"f", 182, 0},
	{"g", 1, 0},
	{"h", 205, 127},
	{"i", 1.21 * 1024, 1.19 * 1024},
	{"j", 25, 25},
	{"k", 21.22 * 1024, 20.95 * 1024},
	{"l", 32.75 * 1024, 32.23 * 1024},
	{"m", 18, 18},
	{"n", 473.30 * 1024, 437.18 * 1024},
	{"o", 27, 0},
	{"p", 49.09 * 1024, 48.75 * 1024},
	{"q", 18.85 * 1024, 18.69 * 1024},
	{"r", 4.52 * 1024, 4.10 * 1024},
	{"s", 152.12 * 1024, 147.25 * 1024},
	{"t", 2.280 * 1024 * 1024, 1.653 * 1024 * 1024},
	{"x", 4.68 * 1024, 4.18 * 1024},
	{"y", 1.16 * 1024, 1.14 * 1024},
}

func Perm(a []int, f func([]int)) {
	perm(a, f, 0)
}

// Permute the values at index i to len(a)-1.
func perm(a []int, f func([]int), i int) {
	if i > len(a) {
		f(a)
		return
	}
	perm(a, f, i+1)
	for j := i + 1; j < len(a); j++ {
		a[i], a[j] = a[j], a[i]
		perm(a, f, i+1)
		a[i], a[j] = a[j], a[i]
	}
}

func loss(a []int) float64 {
	total := float64(0)
	for i := range a {
		h := table[a[i]]
		u := h.Size - h.Dup

		s := float64(0)
		for j := i; j < len(a); j++ {
			s += table[a[j]].Size
		}
		total += u * s
	}
	return total
}

func TestMinLossAllPerm(t *testing.T) {
	a := make([]int, len(table))
	for i := range a {
		a[i] = i
	}

	min := math.MaxFloat64
	best := make([]int, len(table))

	Perm(a, func(a []int) {
		v := loss(a)
		if v < min {
			min = v
			copy(best, a)
		}
	})

	m := float64(1024 * 1024)
	for _, pos := range best {
		fmt.Println(table[pos].Table, fs.SizeSuffix(table[pos].Size*m), fs.SizeSuffix(table[pos].Dup*m), fs.SizeSuffix((table[pos].Size-table[pos].Dup)*m))
	}

	fmt.Printf("%d", int64(loss(best)))
}

type optimizer struct {
	rand *rand.Rand
	beta float64
	loss func([]int) float64
	plot func(int, float64)
}

func newOptimizer(beta float64, loss func([]int) float64, plot func(int, float64)) *optimizer {
	return &optimizer{
		rand: rand.New(rand.NewSource(timeutc.Now().Unix())),
		beta: beta,
		loss: loss,
		plot: plot,
	}
}

func (o *optimizer) metropolis(perm []int, n int) []int {
	cur := perm
	curv := o.loss(cur)
	prop := make([]int, len(cur))

	for i := 0; i < n; i++ {
		o.propose(prop, cur)
		v := loss(prop)

		var p float64
		if v < curv {
			p = 1
		} else {
			p = math.Exp(-o.beta * (v - curv))
		}
		if o.rand.Float64() <= p {
			copy(cur, prop)
			curv = v
		}
		if o.plot != nil {
			o.plot(i, curv)
		}
	}

	return cur
}

func (o *optimizer) propose(dst, src []int) {
	copy(dst, src)
	o.randomSwap(dst)
	o.randomSwap(dst)
}

func (o *optimizer) randomSwap(a []int) {
	n := int32(len(a))
	i := o.rand.Int31n(n)
	j := o.rand.Int31n(n)
	a[i], a[j] = a[j], a[i]
}

func TestOptimizer(t *testing.T) {
	var xy plotter.XYs

	f := func(step int, value float64) {
		xy = append(xy, plotter.XY{
			X: float64(step),
			Y: value,
		})
	}
	a := rand.Perm(len(table))
	o := newOptimizer(0.00001, loss, f)
	best := o.metropolis(a, 4*len(table))

	m := float64(1024 * 1024)
	for _, pos := range best {
		fmt.Printf("%s\t\t\t%s\t%s\n", table[pos].Table, fs.SizeSuffix((table[pos].Size-table[pos].Dup)*m), fs.SizeSuffix(table[pos].Size*m))
	}
	fmt.Printf("%d", int64(loss(best)))

	p := plot.New()
	p.Title.Text = "Plotutil example"
	p.X.Label.Text = "X"
	p.Y.Label.Text = "Y"
	if err := plotutil.AddLinePoints(p, xy); err != nil {
		panic(err)
	}

	// Save the plot to a PNG file.
	if err := p.Save(4*vg.Inch, 4*vg.Inch, "points.png"); err != nil {
		panic(err)
	}
}

//func (o *optimizer) normalDensity(x float64) float64 {
//	const invSqrt2pi = 0.3989422804014327
//	a := x
//	if o.sigma != 1 {
//		a /= o.sigma
//	}
//	return invSqrt2pi / o.sigma * math.Exp(-0.5*a*a)
//}
//
//func TestNormalDensity(t *testing.T) {
//	o := newOptimizer(nil)
//	o.sigma = 2.23
//
//	for v := -5; v < 5; v++ {
//		t.Log(v, o.normalDensity(float64(v)))
//	}
//}

// https://mathematica.stackexchange.com/questions/201333/optimization-of-function-taking-a-permutation
// https://www.coursera.org/lecture/mcmc/the-metropolis-algorithm-for-sampling-eOI1B
// https://towardsdatascience.com/bayesian-statistics-metropolis-hastings-from-scratch-in-python-c3b10cc4382d
// https://towardsdatascience.com/mcmc-intuition-for-everyone-5ae79fff22b1
// https://github.com/gonum/gonum/blob/a382475b2cbd511cea23c7935de6b7a2d15504d7/stat/sampleuv/sample.go#L338
```

**Comment by mmatczuk** [comment]
While this is mathematically correct it assumes that each byte is equally important.
This is not the case if we take the snapshot and compaction mechanics into the picture.
Snapshot creates a hard links to a sstable files that are subject to compaction.
The problem only exists when compaction decides to remove a file that is referenced by a hard link in a snapshot.

For example if we have a table that can be almost fully deduplicated it means two things:
- uploading only few bytes would free up additional disk space, we have very good effort to result ratio (this is what we optimize at the moment with the loss function)
- most likely the table is stagnant and will not be compacted so the risk of keeping the links is low

The other approach to the problem would be to upload first the most active tables.
That is tables that have the most new data because the risk of compaction is the biggest in those tables.
In that context we should upload the newest files first.
If a file survived many compactions chances are it would survive another.
